### PR TITLE
Add basic runtime benchmark infrastructure

### PR DESCRIPTION
**Pull request by Kobzol**
This PR contains basic `collector` infrastructure for runtime benchmarks. It implements the ideas described in https://hackmd.io/j19WFamVQlmQMRigF9wrFw.

There are no DB or UI changes yet, first I think that we should only support the benchmarks locally, for experimentation. Later I will add support for storing the results into the DB.

Currently the design looks like this:
- A new crate was added to the `collector` directory. It's called `benchlib` and it contains a small library for defining and measuring microbenchmarks (using perf. counters with `perf_event_open` and walltime currently). It's used by `collector` to read benchmark results (transferred with JSON currently) and by the runtime benchmarks to define what should be measured.
- There is a directory `collector/runtime-benchmarks` that will contain runtime benchmarks that will use `benchlib` to define a set of related benchmarks. The benchmarks are not executed right away, the `benchlib` instead decides what to do once it parses CLI arguments (it could have commands e.g. to list available benchmarks and their core requirements, so that when you actually execute the benchmark, you can select to how many cores you could pin it).
- A single runtime benchmark was added that measures hashmap insertions.
- A new command called `bench_runtime_local` was added to `collector`. It takes a path to `rustc`, uses it to compile the runtime benchmarks and then executes them. Eventually, it might make sense to instead create something like `bench_local comptime --scenarios ...` and `bench_local runtime ...`. A new `BenchNext` command shouldn't be needed, since it should decide from the `next_artifact` API call if it should also run runtime benchmarks.

Best reviewed commit by commit.

### CODE REVIEW

**Comment by anp** [comment]
This is awesome! After this last RustConf I've been thinking about spending time on this kind of thing again, let me know if I can help! I'm still waiting on some new Linux hardware to be able to use perf_event_open locally but I'll be able to help with results/UI/regression detection/porting benchmarks/etc soon.

**Comment by Kobzol** [comment]
I think that we first have to get the basic infrastructure in place (code layout, measurements, DB, etc.), I'm not sure how much that is "parallelizable". But once that's in place, we will need to populate the benchmark suite and possibly build some new UI for that, and then it will definitely be good to cooperate :) It might still be a way off though, the infrastructure will probably go through a lot of iterations.

File: collector/lolperf/benchlib/src/measure.rs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+pub fn create_group() -> anyhow::Result<Group> {
+    match Group::new() {
+        Ok(group) => Ok(group),
+        Err(error) => {
+            let path = "/proc/sys/kernel/perf_event_paranoid";
+            let level = std::fs::read_to_string(path).unwrap_or_else(|_| "unknown".to_string());
+            let level = level.trim();
+            Err(anyhow::anyhow!(
+                "Cannot create perf_event group ({:?}). Current value of {} is {}. Try lowering it.",
```
Comment (nnethercote) [inline collector/lolperf/benchlib/src/measure.rs:59]: More info about what command to run to change it?

**Review by nnethercote — COMMENTED** [review]
I've had a quick look, it all seems quite plausible.

Why do the benchmarks appear twice? E.g. `collector/lolperf/benchmarks/nbody/src/nbody.rs` and `runtime/benchmarks/nbody/src/nbody.rs`. Same story with other source files like `measure.rs`.

I have a strong preference for the structure of the runtime benchmarks and the compile time benchmarks to be similar. E.g. compile time benchmarks are in `collector/benchmarks/`, so the runtime ones should be in a related spot at the same level of nesting. This will make it easier to navigate. If this requires renaming some existing things, that's fine. (Likewise with the subcommand names and structures, as you suggested.)

File: collector/lolperf/benchlib/src/measure.rs (hunk 1 above)
Comment (Kobzol) [reply collector/lolperf/benchlib/src/measure.rs:59]: Good point. It's not really a specific command, you just have to write the value into the given file path (with .e.g. `echo`). I'll add the hint.

**Comment by Kobzol** [comment]
I also want to make the structure similar to comptime benchmarks, I just put it into a separate directory for now to better separate it from the rest of the code. There's also one additional thing instead of just a directory of benchmarks, namely the benchmarking library. I'm fine with changing the file/directory layout in any way we find better.

**Comment by Kobzol** [comment]
> Why do the benchmarks appear twice? E.g. collector/lolperf/benchmarks/nbody/src/nbody.rs and runtime/benchmarks/nbody/src/nbody.rs. Same story with other source files like measure.rs.

Ooos, sorry, I accidentally committed another directory where I have stored it previously. And I was wondering why the diff is so large.. :sweat_smile:

**Comment by Kobzol** [comment]
I got some nasty diffs there because of a botched directory move, now it should be fine.

**Comment by Kobzol** [comment]
@rylev Can you take a look please? :)

File: collector/benchlib/src/measure/perf_counter/unix.rs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+/// (Micro-)benchmarks a single function.
+pub fn benchmark_function<F: FnOnce() -> R, R>(
+    name: &'static str,
+    func: F,
+) -> anyhow::Result<BenchmarkResult> {
+    let mut group = create_group()?;
+    let counters = prepare_counters(&mut group)?;
+
+    // FIXME: don't run perf counters and time measurements together, run the benchmark twice instead
```
Comment (nnethercote) [inline collector/benchlib/src/measure/perf_counter/unix.rs:23]: Add `.` at end of sentence. Likewise for some comments below.

File: collector/benchlib/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,18 @@
+[package]
+name = "benchlib"
+version = "0.1.0"
+edition = "2021"
+description = "Library for defining and measuring benchmarks of Rust code"
```
Comment (nnethercote) [inline collector/benchlib/Cargo.toml:5]: This is the only comment explaining what benchlib is. Maybe add a slightly more detailed explanation in one of the `.rs` files?

File: collector/runtime-benchmarks/hashmap/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,17 @@
+use benchlib;
+use benchlib::benchmark::BenchmarkSuite;
+
+fn main() {
+    let mut suite = BenchmarkSuite::new();
+
+    suite.register("hashmap-insert-10k", || {
+        let mut map =
+            hashbrown::HashMap::with_capacity_and_hasher(10000, fxhash::FxBuildHasher::default());
```
Comment (nnethercote) [inline collector/runtime-benchmarks/hashmap/src/main.rs:9]: I don't like this benchmark much, I'm hoping the suite will end up much less microbenchmark-y than this. But it's fine as a starting point.

**Review by nnethercote — APPROVED** [review]
I'm annoyed by the inconsistency between the `runtime-benchmarks` and `compile-benchmarks` directory names, i.e. the use of `time` in one but not the other. `compile-time-benchmarks` would be more consistent, but longer. (There's still the inconsistency between `runtime` being a single word and `compile time` being two words but that's a pre-existing bug in English :grin:)

`run-benchmarks` would also be more consistent, but then it sounds like a verb phrase, which is no good.

I don't know, maybe it's good enough as is.

Other than that, and a few minor comments above, this looks good to me.

File: collector/benchlib/Cargo.toml (hunk 3 above)
Comment (Kobzol) [reply collector/benchlib/Cargo.toml:5]: Good idea, I added a comment to the `lib.rs` file.

**Comment by Kobzol** [comment]
We can also name it `comptime-benchmarks`? But I think that `compile` vs `runtime` is obvious enough.

**Comment by nnethercote** [comment]
Ok, `compile-benchmarks` is good enough.

**Comment by Kobzol** [comment]
I'm going to go ahead and merge it so that we can start experimenting with it.

**Pull request merged**